proposal.rst

```diff
@@ -150,14 +146,14 @@ Motivations
 
 **Why have you chosen this particular project?**
 
-FILLME
+This project has many interesting subtasks which I think I will enjoy doing. Also, an open source project is a great way to improve programming skills. There is a lot of scope to think something different and implement it such that to make work. 
 
 **Who will benefit from your project and in what ways?**
 
 .. For example, think about the likely user-base, what they currently have to
 .. do and how your project will improve things for them.
 
-FILLME
+After the implementation of this project, the omindex indexer will use external extractor programs which are available as shared libraries. These libraries will be loaded dynamically during the runtime which will avoid the expense of running other external filters and hence, will speed up indexing as dynamic linking is noticeably faster than static linking. Hence, users of omega will benefit as the search time will be reduced.  
```

Comment:
I thought the external filter programs are statically linked, in contrast to what the libraries we would be using because "using these instead would avoid the overhead of running an external filter and so speed up indexing" as given in the idea description. If not, then what factors are affecting on the speeding up of indexing in the case of libraries?

Reply:
The overhead of running an external filter is mostly in executing the new program, which will involve replacing the mapped memory space, loading the new program code, doing any relocation of internal pointers, resolving and runtime linking dynamic libraries, initialising anything that needs it, and probably a bunch of other things I've forgotten — all before it gets to run any code.

By contrast, we want to continue running an existing program, so we're just paying the cost of resolving and runtime linking dynamic libraries, and a small amount of initialisation. (If that's still too high a cost, I guess we could deal with the dynamic libraries in the parent process, and carry them across the `fork()` call.)

Static linking has essentially nothing to do with this; that's a build-time alternative to dynamic linking, and means that the executable will contain all the library functions it needs. Dynamic linking by contrast happens at runtime, typically when you execute a new program (as above), although you can do it under command of the program using calls like `dlopen()`.

It may be worth you reading up on this, because understanding dynamic linking is important to using `dlopen()` effectively.

Reply:
Now I understood it. Although I know the process of dynamic linking, I know very less about doing it under command using calls like dlopen(). I will read about it and change accordingly. Thank you.